src/dune_rules/alias_rec.ml

```diff
@@ -1,23 +1,129 @@
 open Import
 
-include Action_builder.Alias_rec (struct
-  module Map_reduce =
-    Source_tree.Dir.Make_map_reduce
-      (Action_builder)
-      (Action_builder.Alias_status)
-
-  let traverse dir ~f =
-    let ctx_name, src_dir = Path.Build.extract_build_context_exn dir in
-    let build_dir = Context_name.build_dir (Context_name.of_string ctx_name) in
-    let f dir =
-      let path =
-        Path.Build.append_source build_dir (Source_tree.Dir.path dir)
+module Alias_status = struct
+  module T = struct
+    type t =
+      | Defined
+      | Not_defined
+
+    let empty : t = Not_defined
+
+    let combine : t -> t -> t =
+     fun x y ->
+      match (x, y) with
+      | _, Defined | Defined, _ -> Defined
+      | Not_defined, Not_defined -> Not_defined
+  end
+
+  include T
+  include Monoid.Make (T)
+end
+
+module In_melange_target_dir = struct
+  module Lookup_alias = struct
+    type result =
+      { alias_exists : Alias_status.t
+      ; allowed_subdirs : Filename.Set.t
+      }
+
+    let of_dir_set ~exists dirs =
+      let allowed_subdirs =
+        match Dir_set.toplevel_subdirs dirs with
+        | Infinite -> Filename.Set.empty
+        | Finite sub_dirs -> sub_dirs
       in
-      f ~path
+      { alias_exists = exists; allowed_subdirs }
+  end
+
+  let dep_on_alias_if_exists alias =
+    let open Action_builder.O in
+    Action_builder.of_memo
+      (Load_rules.load_dir ~dir:(Path.build (Alias.dir alias)))
+    >>= function
+    | Source _ | External _ ->
+      Code_error.raise "Alias in a non-build dir"
+        [ ("alias", Alias.to_dyn alias) ]
+    | Build { aliases; allowed_subdirs; rules_here = _ } -> (
+      match Alias.Name.Map.mem aliases (Alias.name alias) with
+      | false ->
+        Action_builder.return
+          (Lookup_alias.of_dir_set ~exists:Not_defined allowed_subdirs)
+      | true ->
+        Action_builder.alias alias
+        >>> Action_builder.return
+              (Lookup_alias.of_dir_set ~exists:Defined allowed_subdirs))
+    | Build_under_directory_target _ ->
+      Action_builder.return
+        { Lookup_alias.alias_exists = Not_defined
+        ; allowed_subdirs = Filename.Set.empty
+        }
+
+  let dep_on_alias_rec name dir =
+    let rec map_reduce dir ~f =
+      let open Action_builder.O in
+      let* { Lookup_alias.alias_exists; allowed_subdirs } = f dir in
+      Action_builder.List.fold_left (String.Set.to_list allowed_subdirs)
+        ~init:alias_exists ~f:(fun alias_exists s ->
+          let+ alias_exists' = map_reduce (Path.Build.relative dir s) ~f in
+          Alias_status.combine alias_exists alias_exists')
     in
+    map_reduce dir ~f:(fun dir -> dep_on_alias_if_exists (Alias.make ~dir name))
+end
+
+let rec map_reduce ~build_dir ~name t ~f =
```

Comment:
Because we need to visit directories that `Source_tree` doesn't find.

Reply:
Because we need to visit the build path, right?

```
    let build_path =
      Path.Build.append_source build_dir (Source_tree.Dir.path t)
```

Is it possible to use the old traversal and just "traverse" the extra build directory in `f`?

Reply:
let me try to make that work.